### ISSUE TITLE
[Question] How does Producer/Consumer reliability work?

### ISSUE THREAD

**Issue by dandago** [confluentinc/confluent-kafka-dotnet#404]
I was trying out some of the samples of the Confluent Kafka .NET client and I noticed neither the consumer nor the producer throws an exception at any stage in the code if the broker is not even running.

I presume this is by design and that the producer/consumer would automatically resume its operation when the broker becomes available again, but I was wondering whether you could point me to any resource explaining how all of this works. I also think it would be useful to have some kind of notification when the producer/consumer can't connect to a broker, to be aware of any outages.

**Comment by dandago** [comment]
Using Confluent.Kafka 0.11.3.

**Comment by comdw** [comment]
Both the Consumer and Producer have an OnError event handler you can bind to that will fire for issues with the broker (and other issues).  There are a bunch of [known errors](https://github.com/confluentinc/confluent-kafka-dotnet/blob/master/src/Confluent.Kafka/ErrorCode.cs) you could have your application check against or you could simply hook these events into your log framework.

```
        /// <summary>
        ///     Raised on critical errors, e.g. connection failures or all 
        ///     brokers down. Note that the client will try to automatically 
        ///     recover from errors - these errors should be seen as 
        ///     informational rather than catastrophic
        /// </summary>
        public event EventHandler<Error> OnError;
```

**Comment by mhowlett** [comment]
yep, thanks @comdw. specifics around this really need documenting better. taking this as a +1. it's in the priority list.

**Comment by jbouzaglou** [comment]
What would really be useful regarding those errors is an excerpt about how is a client app supposed to handle them. For example if you get a "broker connection closed", what should the client do? etc..

**Comment by edenhill** [comment]
https://github.com/edenhill/librdkafka/blob/master/INTRODUCTION.md#message-reliability

**Comment by mhowlett** [comment]
i'm trying to understand the need for `OnError` myself. When would we ever want to act on an event provided by this method? Why is it needed in addition to `OnLog`? I understand that `OnLog` can be called sporadically from any librdkafka thread, where as `OnError` happens as a side-effect of poll (on the calling thread), but I don't understand the use-case of supplying error events to the user separated from log messages if all the user will ever want to do is log that the error occured.

**Comment by mhowlett** [comment]
changing label to enhancement. to answer my commentary above, error messages are indeed intended to be events the user may want to react to. however, to my knowledge there isn't a clear list of which error codes may be emitted anywhere (need to refer to the librdkafka code), or guidance on what the scenarios what actions may be appropriate as a response. enhancement is to do what @jbouzaglou suggests.
